irpc-iroh/examples/derive.rs

```diff
@@ -1,228 +1,204 @@
-                endpoint,
-                addr,
-                b"RPC-Storage".to_vec(),
-            )),
-        })
+    pub struct StorageApi {
+        inner: Client<StorageMessage, StorageProtocol, StorageService>,
     }
 
-    pub fn listen(&self, endpoint: iroh::Endpoint) -> anyhow::Result<AbortOnDropHandle<()>> {
-        let local = self
-            .inner
-            .local()
-            .context("can not listen on remote service")?;
-        let handler: Handler<StorageProtocol> = Arc::new(move |msg, _, tx| {
-            let local = local.clone();
-            Box::pin(match msg {
-                StorageProtocol::Get(msg) => local.send((msg, tx)),
-                StorageProtocol::Set(msg) => local.send((msg, tx)),
-                StorageProtocol::List(msg) => local.send((msg, tx)),
-            })
-        });
-        Ok(AbortOnDropHandle::new(task::spawn(listen(
-            endpoint, handler,
-        ))))
-    }
+    impl StorageApi {
+        pub const ALPN: &[u8] = b"irpc-iroh/derive-demo/0";
 
-    pub async fn get(&self, key: String) -> anyhow::Result<oneshot::Receiver<Option<String>>> {
-        let msg = Get { key };
-        match self.inner.request().await? {
-            Request::Local(request) => {
-                let (tx, rx) = oneshot::channel();
-                request.send((msg, tx)).await?;
-                Ok(rx)
-            }
-            Request::Remote(request) => {
-                let (_tx, rx) = request.write(msg).await?;
-                Ok(rx.into())
-            }
+        pub fn spawn() -> Self {
+            StorageActor::spawn()
         }
-    }
 
-    pub async fn list(&self) -> anyhow::Result<spsc::Receiver<String>> {
-        let msg = List;
-        match self.inner.request().await? {
-            Request::Local(request) => {
-                let (tx, rx) = spsc::channel(10);
-                request.send((msg, tx)).await?;
-                Ok(rx)
-            }
-            Request::Remote(request) => {
-                let (_tx, rx) = request.write(msg).await?;
-                Ok(rx.into())
-            }
+        pub fn connect(endpoint: Endpoint, addr: impl Into<iroh::NodeAddr>) -> Result<StorageApi> {
+            let conn = IrohRemoteConnection::new(endpoint, addr.into(), Self::ALPN.to_vec());
+            Ok(StorageApi {
+                inner: Client::boxed(conn),
+            })
         }
-    }
 
-    pub async fn set(&self, key: String, value: String) -> anyhow::Result<oneshot::Receiver<()>> {
-        let msg = Set { key, value };
-        match self.inner.request().await? {
-            Request::Local(request) => {
-                let (tx, rx) = oneshot::channel();
-                request.send((msg, tx)).await?;
-                Ok(rx)
-            }
-            Request::Remote(request) => {
-                let (_tx, rx) = request.write(msg).await?;
-                Ok(rx.into())
-            }
+        pub fn expose(&self) -> Result<impl ProtocolHandler> {
```

Comment:
Not sure about the name expose. This doesn't do anything actively. You expose it once you accept.

Reply:
I was looking at the line
```rust
Router::builder(endpoint)
    .accept(StorageApi::ALPN, api.???)
```
and `expose` sounded intuitive to me here. Open to suggestions though, do you have something in mind?

Reply:
I was just going to go with `protocol_handler` TBH. But I am not great at naming either. So 🤷 , maybe expose is OK. We can always change it.